include/openthread/platform/radio.h

```diff
@@ -856,6 +856,18 @@ bool otPlatRadioIsCoexEnabled(otInstance *aInstance);
  */
 otError otPlatRadioGetCoexMetrics(otInstance *aInstance, otRadioCoexMetrics *aCoexMetrics);
 
+/**
+ * The radio driver calls this method to notify OpenThread that a frame is
+ * about to be transmitted(triggered by interrupt) so that something like
+ * CSL IE could be updated and security would be processed.
+ *
+ * @param[in]   aInstance   The OpenThread instance structure.
+ * @param[in]   aFrame      A pointer to the first byte of the ACK frame.
+ * @param[in]   aLength     The length of the ACK frame.
+ *
+ */
+void otPlatRadioTxStartedInterrupt(otInstance *aInstance, uint8_t *aFrame, uint16_t aLength);
```

Comment:
Hi Abtin, this issue is discussed and I'm planning to implement 'otPlatRadioSetMacFrameCounter' for simulation platform and maybe nrf528xx so that platform layer will maintain the frame counter. In this way, OT function should only need to form the enh-ack and leave the frame counter field empty and platform layer will fill it before send the frame out. Please see the discussion here: https://github.com/openthread/openthread/pull/5057#discussion_r436451913

Reply:
Thanks @brianljt . I read the comment.

> Agree. Discussed with @bukepo, I think we can move forward with the approach for now. I will implement otPlatRadioSetMacFrameCounter and otPlatRadioSetMacKey for simulation radio so that Enh-ACK will have security and frame counter maintain at platform layer which does not rely on OT APIs and can maintain context dependency there.

I see that both you and @bukepo has approved this PR. Are you suggesting to merge this PR as is (with the potential race issue)? I would very highly advise against this. We should not add code that can be incorrect and lead to bug even if we intend to fix/change this later on. 

----

Regarding adding `otPlatRadioSetMacFrameCounter()`, this radio function is currently optional for radio platforms and only applicable when radio platform provides `OT_RADIO_CAPS_TRANSMIT_SEC` capability and wants to handle the frame counter (and therefore OT core would not need to handle it).

What would be the solution related to the enhanced ack (and security counter) for the case where radio platform does not support `OT_RADIO_CAPS_TRANSMIT_SEC` and security frame counter is managed by OT core. I assumed the current PR (i.e., addition of `otPlatRadioTxStartedInterrupt()`) is intended for this case. Am I missing something here?

Reply:
Hi @abtink, as for the approval of this PR, my thought was currently there are no secured Enh-ACK generated so that it is OK to merge this. But we can hold this if the risk is too high until platform support for frame counter is ready.

For `OT_RADIO_CAPS_TRANSMIT_SEC`, both frame counter and CCM* shares the same capability now. Maybe we should separate them with another capability added.

Reply:
BTW, there are some discussion regarding the frame counter use in Enh-ACK in Thread Group. We might need to wait for a concrete conclusion before we move forward with the change.

https://threadgroup.atlassian.net/browse/SPEC-939
https://threadgroup.atlassian.net/browse/SPEC-938

Reply:
Here is my thoughts on Enh-ACK and `OT_RADIO_CAPS_TRANSMIT_SEC`. Like other radio capabilities we have in `SubMac`, a platform can choose to support it or not. If platform does not support one specific radio capability, `SubMac` should be able to provide an alternative.

We can change the API for `TxFrame::GenerateEnhAck` to pass one more parameter `aFrameCounter` in so that this function could be used as a utility function(like `TxFrame::ProcessTransmitAesCcm`) and is context free which means it could be called in interrupt context.

If Radio dose not have the capability `OT_RADIO_CAPS_TRANSMIT_SEC`, the received frame is passed up to `SubMac` in callback `SubMac::HandleReceiveDone`, this is already in OT context. `SubMac` can generate an Enh-ACK, encrypt it and send it out to radio. Increamenting the frame counter in same context avoid the racing condition. If radio supports `OT_RADIO_CAPS_TRANSMIT_SEC`, then it still can call the utility function `TxFrame::GenerateEnhAck` to generate an Enh-ACK, but the frame counter is maintained by radio layer so that radio will provide atomicity.

The function `otPlatRadioTxStartedInterrupt` is called when `SubMac` needs to update the CSL IE. We should not update frame counter in this.

@abtink @bukepo @Irving-cl , thoughts?

Reply:
I agree to add a parameter `aFrameCounter` to `TxFrame::GenerateEnhAck`. Then we also need a method to get frame counter at radio layer from `SubMac` right?
Regarding `otPlatRadioTxStartedInterrupt`, after we update the CSL IE, do we need to call `ProcessTransmitAesCcm` again? Shall we call this method at the end of `otPlatRadioTxStartedInterrupt` or call it after `otPlatRadioTxStartedInterrupt`?

Reply:
I think we don't need to get frame counter at radio layer from `SubMac`. If radio supports `OT_RADIO_CAPS_TRANSMIT_SEC`, radio maintains the counter, so it is from radio layer. If not, radio cannot generate Enh-ACK and will leave it to `SubMac` in `SubMac::HandleReceiveDone`.

After CSL IE updated, yes, `ProcessTransmitAesCcm` should be called. I think it could be call at the end of `otPlatRadioTxStartedInterrupt` as long as all the information in header IE has been updated.

Reply:
Passing counter as param in `GenerateEnhAck` sounds reasonable to me. We should ensure caller handles the case where the counter is not needed/used (i.e., ack is not secure). 

> If Radio dose not have the capability OT_RADIO_CAPS_TRANSMIT_SEC, the received frame is passed up to SubMac in callback SubMac::HandleReceiveDone, this is already in OT context. SubMac can generate an Enh-ACK, encrypt it and send it out to radio.

Not sure if it would work reliably (on all platforms). Regular ack is expected to be sent out within a short window of time after frame rx. I expect the same for enh-ack. It would be hard to ensure limited delay allowing context switch and preparing ack from `HandleReceiveDone`.

------

I propose we adopt the following guideline:

_Any platform function that we allow/intend to be called from other than OT OS context (e.g., from an ISR) should not get the `otInstance` and should not access any state or variable within OT. Such a function should only be allowed to use static class methods._ 

(we should not even allow reading of any state/variable, since for it to work correctly we would need to have atomic write/read of all variables in OT). 

Thoughts?

Reply:
I agree with the guideline. As OT does not provide locks on state/variables read/write, OT APIs which need access to OT state/variables should not be called in context other than OT OS context. For an OT API to be used in other context, it should be OT variables independent.

So I think the best place to handle Enh-ACK and frame counter is at radio layer if it supports `OT_RADIO_CAPS_TRANSMIT_SEC`. It would be easier to maintain atomicity and accurate time control there than at OT stack which involves context switch and maybe tasklet scheduling. Otherwise, OT `SubMac` provides an alternative to it as a last resort at the cost of possibility of not meeting the criterial to send out ACK within AIFS(192us).

Or do you think we should force Radio to support `OT_RADIO_CAPS_TRANSMIT_SEC` if build version is OT1.2?

BTW, there are some discussion in the Thread group regarding the frame counter used in Enh-ACK(Check this comment, https://threadgroup.atlassian.net/browse/SPEC-939?focusedCommentId=48001). It tries to use the frame counter in received frame with some variance in nonce generation to avoid reusing the same key material. This approach is still under discussion. If this approach is adopted by spec, then it will be easier to implement Enh-ACK as it does not rely on the transmit frame counter.